DNA methylation analysis with 450K array
========================================================

Contributors: Gloria Li

> This seminar is inspired by one of last year's class project: *Analysis of Gene Expression Omnibus Leukaemia Data from the Illumina HumanMethylation450 Array* by Alice Zhu, Rachel Edgar, Shaun Jackman and Nick Fishbane.    

Illumina HumanMethylation450 Array is a microarray based high throughput platform for methylation profiling on pre-selected 450,000 probes for CpGs across human genome. A great number of datasets have been made publicly available through Gene Expression Omnibus (GEO). Each dataset in GEO has a unique GSE ID that we can use to retrieve it from GEO. There are several Bioconductor packages that are very useful for obtaining data from GEO and microarray analysis. If you haven't done so yet, install and load these packages.   

```{r, eval = FALSE}
source('http://bioconductor.org/biocLite.R')
biocLite('GEOquery')
biocLite('wateRmelon')
biocLite("IlluminaHumanMethylation450k.db")
```
```{r, message=F, warning=F}
library(GEOquery)
library(wateRmelon)
library(IlluminaHumanMethylation450k.db)
```

## Explore 450k methylation array data
In this seminar, we are going to perform differential methylation analysis for Acute lymphoblastic leukemia (ALL) cases and healty B cells as control group. The datasets we are going to use are: GSE39141 which contains 29 ALL bone marrow samples and 4 healthy B cell samples, and to makeup for the small sample size for healthy cells,  we will also use GSE42865 which contains 9 healthy B cell samples and 7 samlpes with other conditions.    

First, let's retrieve our datasets from GEO with `getGEO` from `GEOquery` package. This may take a while.   

```{r fetchGEO, message = F}
GSE39141 <- getGEO('GSE39141') 
show(GSE39141) ## 33 samples (29 ALL and 4 healthy B cells)
GSE42865 <- getGEO('GSE42865') 
show(GSE42865) ## 16 samples (9 healthy cells B cells and 7 other cells)
```

As you can see, the datasets we got are large lists including Beta values as well as metadata. We can extract these information separately. Then we can take a subset of GSE42865 that are healthy cells.    

```{r extract}
# Extract expression matrices (turn into data frames at once) 
ALL.dat <- as.data.frame(exprs(GSE39141[[1]]))
CTRL.dat <- as.data.frame(exprs(GSE42865[[1]]))

# Obtain the meta-data for the samples and rename them perhaps?
ALL.meta <- pData(phenoData(GSE39141[[1]]))
CTRL.meta <- pData(phenoData(GSE42865[[1]]))

# create some labels
ALL.meta$Group<- c(rep('ALL', 29), rep('HBC', 4))  ## ALL: Case; HBC: Healthy B Cells

# Subset both meta-data and data for control (healthy) donors
CTRL.meta <- droplevels(subset(CTRL.meta, grepl("Healthy donor", characteristics_ch1.1)))
CTRL.dat <- subset(CTRL.dat, select = as.character(CTRL.meta$geo_accession)) 

# Rename column names
colnames(ALL.dat) <- paste(ALL.meta$Group, gsub("GSM", "", colnames(ALL.dat)), sep = '_')
colnames(CTRL.dat) <- paste('HBC', gsub("GSM", "", colnames(CTRL.dat)), sep = '_')
```

Now, we can do some exploratory analysis of the data, for examples, looking at distribution of Beta values in each sample or each probe. Here is a density plot of average Beta values for probes in the two datasets. See if you get similar results. 
> Hint: you can use `rowMeans` function to calculate the average Beta value for each probe. Don't forget to use `na.rm = T` to exclude NAs. 

```{r exploratory, echo = F}
# density plot
library(ggplot2)
dat.probeMeans <- c(rowMeans(ALL.dat, na.rm = T), rowMeans(CTRL.dat, na.rm = T)) 
plotDat <- data.frame(Beta = dat.probeMeans, Dataset = rep(c('ALL', 'CTRL'), each = nrow(ALL.dat)))
(probeAvg <- ggplot(data = plotDat, aes(x = Beta, col = Dataset)) +
   geom_density() + 
   ggtitle("Density of Beta values for 2 arrays") + 
   xlab("Beta") + 
   ylab("Density") + 
   theme_bw()
)
```

## Normalization
You can see the distribution of Beta values are not the same between different experiments. That is why it is important to normailze the data before proceed. The Bioconductor package `wateRmelon` offers 15 methods for 450K methylation array normalization. If you are interested in comparing these methods, check out [this paper](http://www.biomedcentral.com/1471-2164/14/293). For the sake of convenience, we will just use the simplest method, `betaqn`, which performs quantile normailzation on the beta values.        

```{r normalization}
# combine data from two experiments into one matrix, each column represents beta values of one sample
beta.matrix <- as.matrix(cbind(ALL.dat, CTRL.dat))
str(beta.matrix, max.level = 0)

# quantile normalization
beta.norm <- betaqn(beta.matrix)
```
```{r postNorm, echo = F}
dat.probeMeans <- c(rowMeans(beta.norm[,1:33], na.rm = T), rowMeans(beta.norm[,34:42], na.rm = T)) 
plotNorm <- rbind(data.frame(plotDat, Norm = "Before"), data.frame(Beta = dat.probeMeans, Dataset = rep(c('ALL', 'CTRL'), each = nrow(ALL.dat)), Norm = "After"))
plotNorm$Norm <- factor(plotNorm$Norm, levels = c("Before", "After"))
(probeAvgNorm <- ggplot(data = plotNorm, aes(x = Beta, col = Dataset)) +
   geom_density() + 
   facet_grid(Norm ~ .) + 
   ggtitle("Density of Beta values before and after normalization") + 
   xlab("Beta") + 
   ylab("Density") + 
   theme_bw()
)
```

## M values
As you can see, after normalization, beta values from different experiment have more similar distributions. However, this distribution is sort of a binary distribution between 0 and 1. This type of distribution does not perform well under linear regression models. A common solution is to apply a logit transformation on the data to convert it to a continuous variable that span $(-\infty, \infty)$, i.e. compute the M value. `logit` function in `gtools` package provides a simple way for this conversion. 
$$M = log({Beta\over 1-Beta})$$

```{r Beta2M}
library(gtools)
M.norm <- logit(beta.norm)
```

## CpG Islands
Now we can go ahead with differential methylation analysis. You can perform this analysis on each probe, but in this seminar, we will aggregate the probes into CpG Islands (CGIs), i.e. the CpG dense regions of the genome, and then detect differential methylated CGIs. The biological function of CGIs is better studied, so the interpretation of our results will be easier if we focus on CGIs. Conveniently, the Bioconductor package `IlluminaHumanMethylation450k.db` provides all sorts of annotation information for the 450K methylation array including the association of each probe to CGIs. We will extract this information and use the mean of M values of a specific CGI to represent its methylation level. Check if you get the same boxplot for CGI M values.    

```{r aggregate}
# Extracting probe ID to CpG islands association
cginame <- as.data.frame(IlluminaHumanMethylation450kCPGINAME)
colnames(cginame) <- c('Probe_ID', 'cginame')
rownames(cginame) <- cginame$Probe_ID
length(levels(factor(cginame$cginame)))   # No. of CGIs

# restrict probes to those within CGIs
beta.inCGI <- beta.norm[cginame$Probe_ID,]
M.inCGI <- M.norm[cginame$Probe_ID,]
nrow(M.inCGI)  # No. of probes within CGIs

# aggregate probes to CGIs
beta.CGI <- aggregate(beta.inCGI, by = list(cginame$cginame), mean, na.rm = T)
rownames(beta.CGI) <- beta.CGI[,1]
beta.CGI <- beta.CGI[,2:ncol(beta.CGI)]
str(beta.CGI, max.level=0)
M.CGI <- aggregate(M.inCGI, by = list(cginame$cginame), mean, na.rm = T)
rownames(M.CGI) <- M.CGI[,1]
M.CGI <- M.CGI[,2:ncol(M.CGI)]
str(M.CGI, max.level = 0)
```
```{r M.CGI.boxplot, echo=F, warning=F}
# check the distribution of CGI M values with boxplot
library(reshape2)
M.CGI.tall <- melt(t(M.CGI), value.name = 'M', varnames = c('Sample', 'CGI'))
M.CGI.tall$Group <- gsub("_[0-9]+", "", M.CGI.tall$Sample)
(M.boxplot <- ggplot(data = M.CGI.tall, aes(Sample, M, fill = factor(Group))) + 
   geom_boxplot() + 
   ggtitle("Distribution of CGI M values") + 
   xlab("Samples") + 
   ylab("M values") + 
   theme_bw() + 
   scale_x_discrete(labels = NULL))
```

## Differential methylation analysis with limma
Next, we can use linear model to identify differential methylated CGIs with `limma`. You are already familiar with this part.    

```{r limma}
library(limma)
design <- data.frame(Group = relevel(factor(gsub("_[0-9]+", "", colnames(M.CGI))), ref = "HBC"), row.names = colnames(M.CGI))
str(design)
(DesMat <- model.matrix(~Group, design))
DMRfit <- lmFit(M.CGI, DesMat)
DMRfitEb <- eBayes(DMRfit)
cutoff <- 1e-2
DMR <- topTable(DMRfitEb, coef = 'GroupALL', number = Inf, p.value = cutoff)
head(DMR)   # top hits 
```

So using a cutoff of BH adjusted p-value equals 0.01, we identified `r nrow(DMR)` CGIs that are differentially methylated between ALL and control group. Now we can make some plots to check these hits. 

```{r plotHits, message=F}
# heatmap of beta values of top 100 hits 
library(gplots)
DMR100 <- topTable(DMRfitEb, coef = 'GroupALL', number = 100)
DMR.CGI <- t(as.matrix(subset(beta.CGI, rownames(beta.CGI) %in% rownames(DMR100))))
str(DMR.CGI, max.level = 0)
col <- c(rep("darkgoldenrod1", times = 29), rep("forestgreen", times = 13))
par(mai = rep(0.5, 4))
heatmap.2(DMR.CGI, col = redblue(256), RowSideColors = col, density.info = "none", trace = "none", Rowv = T, Colv = T, labCol = F, labRow = F, dendrogram="row", margins = c(1, 5), keysize = 1)
legend("topright", c("ALL", "HBC"), col = c("darkgoldenrod1", "forestgreen"), pch = 15)

# stripplot of beta values of probes within top 5 CGI hits
DMR5 <- topTable(DMRfitEb, coef = 'GroupALL', number = 5)
beta.DMR5probe <- beta.inCGI[cginame[rownames(beta.inCGI),]$cginame %in% rownames(DMR5),]
beta.DMR5probe.tall <- melt(beta.DMR5probe, value.name = 'M', varnames = c('Probe_ID', 'Sample'))
beta.DMR5probe.tall$Group <- factor(gsub("_[0-9]+", "", beta.DMR5probe.tall$Sample))
beta.DMR5probe.tall$CGI <- factor(cginame[as.character(beta.DMR5probe.tall$Probe_ID),]$cginame)
(beta.DMR5.stripplot <- ggplot(data = beta.DMR5probe.tall, aes(x = Group, y = M, color = Group)) + 
   geom_point(position = position_jitter(width = 0.05), na.rm = T) + 
   facet_grid(. ~ CGI) + 
   ggtitle("Probe beta values within top 5 DM CGIs") + 
   xlab("Group") + 
   ylab("beta") + 
   theme_bw())

# plot location of differential methylated probes along each chromosome
chrlen <- unlist(as.list(IlluminaHumanMethylation450kCHRLENGTHS)[c(as.character(1:22), "X")])   # get the length of chromosome 1-22 and X
chrlen <- data.frame(chr = factor(names(chrlen)), length = chrlen)
chr <- IlluminaHumanMethylation450kCHR           # get the chromosome of each probe
chr <- unlist(as.list(chr[mappedkeys(chr)]))     # get the probe identifiers that are mapped to chromosome
coord <- IlluminaHumanMethylation450kCPGCOORDINATE      # get chromosome coordinate of each probe
coord <- unlist(as.list(coord[mappedkeys(coord)]))      # get the probe identifiers that are mapped to coordinate
coord <- data.frame(chr = chr[intersect(names(chr), names(coord))], coord = coord[intersect(names(chr), names(coord))])
coordDMRprobe <- droplevels(na.omit(coord[cginame[cginame$cginame %in% rownames(DMR),]$Probe_ID,])) # coordinates of probes in DM CGIs
(coord.plot <- ggplot(data = coordDMRprobe) + 
   geom_linerange(aes(factor(chr, levels = c("X", as.character(22:1))), ymin = 0, ymax = length), data = chrlen, alpha = 0.5) + 
   geom_point(aes(x = factor(chr, levels = c("X", as.character(22:1))), y = coord), position = position_jitter(width = 0.03), na.rm = T) + 
   ggtitle("DMR positions on chromosomes") + 
   ylab("Position of DMRs") +
   xlab("chr") +
   coord_flip() + 
   theme_bw())
```

## Interpretation and functional enrichment analysis
The next step in the analysis pipeline would be to interpret our result by associating these differential methylated regions (DMRs) with biological features. DNA methylation at different genomic regions, i.e. promoter regions, enhancers, genebody etc., has different impact on gene transcription. A common practice in DMR analysis is to separate DMR associated with different types of genomic regions and study them separately. Our key interest is to see how these DMRs affect gene expression level and consequently biological functions. For that purpose, after we separate DMRs into different genomic regions, we will associate them with genes and then biological functions. There are many ways to do this. The `IlluminaHumanMethylation450k.db` package itself contains many annotation information, including associating probes with different types of genomic regions and gene IDs. You can obtain this information similarly to how we extracted chromosome coordinates above. Here is a full list of all objects available in this package. Their names are self-explanatory, but if want more detailed description, check the [package manual document](http://www.bioconductor.org/packages/release/data/annotation/manuals/IlluminaHumanMethylation450k.db/man/IlluminaHumanMethylation450k.db.pdf). 
```{r} 
ls("package:IlluminaHumanMethylation450k.db")
```              

There are a lot of resources that provide association between genes and biological functions. Gene Ontology (GO) is a popular database that provides a controlled vocabulary of terms for describing gene product characteristics and functions including molecular function, biological process and cellular component. Also, there are databases like KEGG and PANTHER that provide association of genes with pathways for pathway analysis, and InterPro and Pfam etc. that provide protein domain analysis. You might noticed from the list above that `IlluminaHumanMethylation450k.db` also provide some of these association analysis like GO, but you would have more control and flexibility if you use other tools more specialized for functional enrichment analysis. There are two online tools that I found very helpful, both extremely easy to use. [DAVID](http://david.abcc.ncifcrf.gov/tools.jsp) takes a list of genes and output annotation and enrichment analysis from all sorts of databases including all the ones we discussed above. The results are available for downloading as Tab deliminated text format. Another more recent tool is [GREAT](http://bejerano.stanford.edu/great/public/html/). It was originally development to associate regulatory regions with genes and biological functions and the required input is a list of coordinates. It also performs functional enrichment analysis from different databases and it can make some fancy plots for output.       

The interpretation of differential methylation analysis results is a very flexible and most often case-specific process. Functional enrichment analysis is just one popular way to approach this but there is really no *gold-standard pipeline* we can follow. Although we won't be able to go into it in this seminar, you should understand that identifying DMRs is really the easy part and digging up the biology behind it and make a story of it would be something you need to spend a lot of time considering.       
